File: _migrate_redis.js
const { createClient } = require('redis');

async function migrate() {
  const r = createClient({ url: 'redis://redis:6379' });
  await r.connect();
  console.log('[OK] Redis conectado');

  // 1. Migrar agenda:msgs:*
  const msgKeys = await r.keys('agenda:msgs:*');
  let migrated = 0, merged = 0;

  for (const key of msgKeys) {
    const phone = key.replace('agenda:msgs:', '');
    const digits = phone.replace(/\D/g, '');

    if (digits.length === 12 && digits.startsWith('55')) {
      const correct = digits.slice(0, 4) + '9' + digits.slice(4);
      const correctKey = `agenda:msgs:${correct}`;

      const oldRaw = await r.get(key);
      const oldMsgs = oldRaw ? JSON.parse(oldRaw) : [];
      const existRaw = await r.get(correctKey);
      const existMsgs = existRaw ? JSON.parse(existRaw) : [];

      if (existMsgs.length > 0) {
        const all = [...existMsgs, ...oldMsgs];
        const seen = new Set();
        const unique = all.filter(m => {
          const sig = `${m.from||''}${m.text||''}${m.ts||0}`;
          if (seen.has(sig)) return false;
          seen.add(sig);
          return true;
        }).sort((a, b) => (a.ts || 0) - (b.ts || 0));
        await r.set(correctKey, JSON.stringify(unique));
        await r.del(key);
        merged++;
        console.log(`  MERGED: ${phone} -> ${correct} (${oldMsgs.length}+${existMsgs.length}=${unique.length})`);
      } else {
        await r.rename(key, correctKey);
        migrated++;
        console.log(`  RENAMED: ${phone} -> ${correct} (${oldMsgs.length} msgs)`);
      }
    }
  }
  console.log(`\n[OK] msgs: ${migrated} renomeadas, ${merged} mescladas`);

  // 2. Migrar agenda:phone2id:*
  const p2iKeys = await r.keys('agenda:phone2id:*');
  let p2i = 0;
  for (const key of p2iKeys) {
    const phone = key.replace('agenda:phone2id:', '');
    const digits = phone.replace(/\D/g, '');
    if (digits.length === 12 && digits.startsWith('55')) {
      const correct = digits.slice(0, 4) + '9' + digits.slice(4);
      const ck = `agenda:phone2id:${correct}`;
      const val = await r.get(key);
      if (!(await r.exists(ck))) {
        await r.rename(key, ck);
      } else {
        await r.del(key);
      }
      p2i++;
    }
  }
  console.log(`[OK] phone2id: ${p2i} migradas`);

  // 3. Migrar agenda:unread:*
  const urKeys = await r.keys('agenda:unread:*');
  let ur = 0;
  for (const key of urKeys) {
    const phone = key.replace('agenda:unread:', '');
    const digits = phone.replace(/\D/g, '');
    if (digits.length === 12 && digits.startsWith('55')) {
      const correct = digits.slice(0, 4) + '9' + digits.slice(4);
      const ck = `agenda:unread:${correct}`;
      const oldVal = parseInt(await r.get(key) || '0');
      const exVal = parseInt(await r.get(ck) || '0');
      await r.set(ck, String(oldVal + exVal));
      await r.del(key);
      ur++;
    }
  }
  console.log(`[OK] unread: ${ur} migradas`);

  // 4. Verificar
  const remaining = (await r.keys('agenda:msgs:55*')).filter(k => {
    const p = k.replace('agenda:msgs:', '');
    return p.length === 12;
  }).length;
  console.log(`\n[RESULTADO] Chaves orfas restantes: ${remaining}`);
  console.log('[SUCESSO] Migracao concluida!');

  await r.quit();
}

migrate().catch(e => { console.error('ERRO:', e.message); process.exit(1); });

File: _ARQUIVO_MORTO/_migrate_redis.py
#!/usr/bin/env python3
"""Migra chaves Redis de 12 digitos (sem 9) para 13 digitos (com 9)"""
import redis
import json

r = redis.from_url("redis://localhost:6379")

# 1. Migrar agenda:msgs:*
keys = r.keys("agenda:msgs:*")
migrated = 0
merged = 0
for key in keys:
    phone = key.decode().replace("agenda:msgs:", "")
    digits = ''.join(c for c in phone if c.isdigit())
    
    if len(digits) == 12 and digits.startswith("55"):
        # Adicionar o 9 apos o DDD (posicao 4)
        correct = digits[:4] + "9" + digits[4:]
        correct_key = f"agenda:msgs:{correct}"
        
        old_msgs = json.loads(r.get(key) or "[]")
        existing = json.loads(r.get(correct_key) or "[]")
        
        if existing:
            # Mesclar: juntar + ordenar por timestamp + deduplicar
            all_msgs = existing + old_msgs
            seen = set()
            unique = []
            for m in all_msgs:
                sig = f"{m.get('from','')}{m.get('text','')}{m.get('ts',0)}"
                if sig not in seen:
                    seen.add(sig)
                    unique.append(m)
            unique.sort(key=lambda x: x.get("ts", 0))
            r.set(correct_key, json.dumps(unique))
            r.delete(key)
            merged += 1
            print(f"  MERGED: {phone} -> {correct} ({len(old_msgs)} + {len(existing)} = {len(unique)} msgs)")
        else:
            # Renomear
            r.rename(key, correct_key)
            migrated += 1
            print(f"  RENAMED: {phone} -> {correct} ({len(old_msgs)} msgs)")

print(f"\n[OK] msgs: {migrated} renomeadas, {merged} mescladas")

# 2. Migrar agenda:phone2id:*
p2i_keys = r.keys("agenda:phone2id:*")
p2i_migrated = 0
for key in p2i_keys:
    phone = key.decode().replace("agenda:phone2id:", "")
    digits = ''.join(c for c in phone if c.isdigit())
    if len(digits) == 12 and digits.startswith("55"):
        correct = digits[:4] + "9" + digits[4:]
        correct_key = f"agenda:phone2id:{correct}"
        if not r.exists(correct_key):
            r.rename(key, correct_key)
        else:
            r.delete(key)
        p2i_migrated += 1

print(f"[OK] phone2id: {p2i_migrated} migradas")

# 3. Migrar agenda:unread:*
ur_keys = r.keys("agenda:unread:*")
ur_migrated = 0
for key in ur_keys:
    phone = key.decode().replace("agenda:unread:", "")
    digits = ''.join(c for c in phone if c.isdigit())
    if len(digits) == 12 and digits.startswith("55"):
        correct = digits[:4] + "9" + digits[4:]
        correct_key = f"agenda:unread:{correct}"
        old_val = r.get(key)
        existing_val = r.get(correct_key)
        if existing_val:
            total = int(old_val or 0) + int(existing_val or 0)
            r.set(correct_key, str(total))
        else:
            r.rename(key, correct_key)
        r.delete(key) if r.exists(key) else None
        ur_migrated += 1

print(f"[OK] unread: {ur_migrated} migradas")

# 4. Verificar: contar chaves de 12 digitos restantes
remaining = 0
for key in r.keys("agenda:msgs:55*"):
    phone = key.decode().replace("agenda:msgs:", "")
    if len(phone) == 12:
        remaining += 1

print(f"\n[RESULTADO] Chaves orfas restantes: {remaining}")
print("[SUCESSO] Migracao concluida!")
